validator/src/state/merkle.rs

```diff
@@ -0,0 +1,894 @@
+/*
+ * Copyright 2018 Intel Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * ------------------------------------------------------------------------------
+ */
+
+use std::collections::BTreeMap;
+use std::collections::HashMap;
+use std::collections::VecDeque;
+use std::io::Cursor;
+
+use cbor;
+use cbor::encoder::{EncodeError, GenericEncoder};
+use cbor::decoder::{DecodeError, GenericDecoder};
+use cbor::value::Value;
+use cbor::value::Key;
+use cbor::value::Bytes;
+use cbor::value::Text;
+
+use crypto::digest::Digest;
+use crypto::sha2::Sha512;
+
+use database::database::DatabaseError;
+use database::lmdb::LmdbDatabase;
+
+const TOKEN_SIZE: usize = 2;
+
+/// Merkle Database
+pub struct MerkleDatabase {
+    root_hash: String,
+    db: LmdbDatabase,
+    root_node: Node,
+}
+
+#[derive(Debug)]
+pub enum MerkleDatabaseError {
+    NotFound(String),
+    DeserializationError(DecodeError),
+    SerializationError(EncodeError),
+    InvalidRecord,
+    DatabaseError(DatabaseError),
+    UnknownError,
+}
+
+impl From<DatabaseError> for MerkleDatabaseError {
+    fn from(err: DatabaseError) -> Self {
+        MerkleDatabaseError::DatabaseError(err)
+    }
+}
+
+impl From<EncodeError> for MerkleDatabaseError {
+    fn from(err: EncodeError) -> Self {
+        MerkleDatabaseError::SerializationError(err)
+    }
+}
+
+impl From<DecodeError> for MerkleDatabaseError {
+    fn from(err: DecodeError) -> Self {
+        MerkleDatabaseError::DeserializationError(err)
+    }
+}
+
+type MerkleResult<T> = Result<T, MerkleDatabaseError>;
+
+impl MerkleDatabase {
+    pub fn new(db: LmdbDatabase, merkle_root: Option<&str>) -> MerkleResult<Self> {
+        let root_hash = match merkle_root {
```

Comment:
It will end up looking like this, since you'll note there is a result/? combo in the None branch:
```
        let root_hash = merkle_root.map(String::from).map(Ok).unwrap_or_else(|| initialize_db(&db))?;
```
Thoughts?